samples/boards/nrf/system_off/src/retained.c

```diff
@@ -0,0 +1,164 @@
+
+/* Number of small blocks */
+#define SMALL_BLOCK_COUNT 8
+
+/* Span of the SRAM area covered by small sections */
+#define SMALL_SECTION_SPAN (SMALL_BLOCK_COUNT		\
+	* SMALL_SECTIONS_PER_BLOCK * SMALL_SECTION_SIZE)
+
+/* Inclusive address of the RAM range covered by large sections */
+#define LARGE_SECTION_BEGIN (SRAM_BEGIN + SMALL_SECTION_SPAN)
+
+/* Size of a controllable RAM section in large blocks */
+#define LARGE_SECTION_SIZE 32768
+
+/* Set or clear RAM retention in SYSTEM_OFF for the provided object.
+ *
+ * @note This only works for nRF52 with the POWER module.  The other
+ * Nordic chips use a different low-level API, which is not currently
+ * used by this function.
+ *
+ * @param ptr pointer to the start of the retainable object
+ *
+ * @param len length of the retainable object
+ *
+ * @param enable true to enable retention, false to clear retention
+ */
+static int ram_range_retain(const void *ptr,
+			    size_t len,
+			    bool enable)
+{
+	uintptr_t addr = (uintptr_t)ptr;
+	uintptr_t addr_end = addr + len;
+
+	/* Error if the provided range is empty or doesn't lie
+	 * entirely within the SRAM address space.
+	 */
+	if ((len == 0U)
+	    || (addr < SRAM_BEGIN)
+	    || (addr > (SRAM_END - len))) {
+		return -EINVAL;
+	}
+
+	/* Iterate over each section covered by the range, setting the
+	 * corresponding RAM OFF retention bit in the parent block.
+	 */
+	do {
+		uintptr_t block_base = SRAM_BEGIN;
+		uint32_t section_size = SMALL_SECTION_SIZE;
+		uint32_t sections_per_block = SMALL_SECTIONS_PER_BLOCK;
+		bool is_large = (addr >= LARGE_SECTION_BEGIN);
+		uint8_t block = 0;
+
+		if (is_large) {
+			block = 8;
+			block_base = LARGE_SECTION_BEGIN;
+			section_size = LARGE_SECTION_SIZE;
+			sections_per_block = UINT32_MAX;
+		}
+
+		uint32_t section = (addr - block_base) / section_size;
+
+		if (section >= sections_per_block) {
+			block += section / sections_per_block;
+			section %= sections_per_block;
+		}
+
+		uint32_t section_mask =
+			(POWER_RAM_POWERSET_S0RETENTION_On
+					 << (section + POWER_RAM_POWERSET_S0RETENTION_Pos));
+
+		if (enable) {
+			nrf_power_rampower_mask_on(NRF_POWER, block, section_mask);
+		} else {
+			nrf_power_rampower_mask_off(NRF_POWER, block, section_mask);
+		}
+
+		/* Move to the first address in the next section. */
+		addr += section_size - (addr % section_size);
+	} while (addr < addr_end);
+
+	return 0;
+}
+
+/* Retained data must be defined in a no-init section to prevent the C
+ * runtime initialization from zeroing it before anybody can see it.
+ */
+__noinit struct retained_data retained;
+
+bool retained_validate(void)
+{
+	/* The residue of a CRC is what you get from the CRC over the
+	 * message catenated with its CRC.  This is the post-final-xor
+	 * residue for CRC-32 (CRC-32/ISO-HDLC) which Zephyr calls
+	 * crc32_ieee.
+	 */
+	const uint32_t residue = 0x2144df1c;
+	uint32_t crc = crc32_ieee((const uint8_t *)&retained,
+				  offsetof(struct retained_data, end));
```

Comment:
Wouldn't using `offsetof(struct retained_data, crc) + sizeof(retained.crc)` here eliminate the need of adding the `end` field and accompanying comment? I think this could be even easier for people to understand what is going on here.
And aren't compilers actually allowed to put some padding between the `crc` and `end` fields?